### PR TITLE
Improvements & fixes for the release pipeline

### LINKED ISSUE #3375: Release pipeline: DSP repo tagged with extra leading "v" character
For the 8.0 Preview 3 release, the commit for the dotnet-security-partners repo was tagged as `vv8.0.0-preview.3.23174.8`. Note the extra `v` which shouldn't be there. The release info from the Pre-release stage indicates a correct value for the tag: `Release tag: v8.0.0-preview.3.23174.8`.

[Link to example build](https://dev.azure.com/dnceng/internal/_build/results?buildId=2155054&view=results) (internal only)

### LINKED ISSUE #3376: Release pipeline: SB tagging pattern not compatible with installer tagging
Because the [SB tagging pattern now uses the runtime version](https://github.com/dotnet/source-build/pull/3321), this is different than the installer repo's tagging pattern (based on the SDK version). This difference causes incompatibility in the [tagging validation](https://github.com/dotnet/source-build/blob/2c32505fda133f677d07ddc66e751e4010c11a10/eng/templates/stages/release.yml#L181-L203) that occurs in the release pipeline. It's comparing the runtime version to the SDK version which is obviously not going to match.

[Example build link](https://dev.azure.com/dnceng/internal/_build/results?buildId=2147525&view=logs&j=19992227-62fb-5b50-4e29-3b72bd33eea1&t=6d20ecd5-829b-5d5e-8095-acf3eddfc277) (internal only)

### LINKED ISSUE #3377: Release pipeline: Incorrect content in generated announcement
The announcement generated by the release pipeline has the following incorrect content:

* `RELEASE_CHANNEL` is not resolved as a variable. It's inserted literally as `RELEASE_CHANNEL` in the content.
* For servicing release:
  * Tag URLs reference https://github.com/dotnet/dotnet instead of https://github.com/dotnet/installer.

For an example of this, you can see the history of the post at https://github.com/dotnet/source-build/discussions/3368.

### LINKED ISSUE #3379: Release pipeline: Incorrect URL to 8.0 preview release notes in announcement
The release announcement created by the release pipeline for the 8.0 Preview 3 release generated a link to the release notes. It referenced this URL which does not exist: https://github.com/dotnet/core/blob/main/release-notes/RELEASE_CHANNEL/8.0.0-preview.3.23174.8/8.0.100-preview.3.23178.7.md. The correct URL is https://github.com/dotnet/core/blob/main/release-notes/8.0/preview/8.0.0-preview.3.md. See the history of the post at https://github.com/dotnet/source-build/discussions/3370.

### LINKED ISSUE #3380: Release pipeline: JSON in release announcement not relevant to VMR
When the release pipeline generates a release announcement, it includes JSON to describe state of the release. One of the fields is `InstallerTag`. This refers to the https://github.com/dotnet/installer repo which is relevant for 8.0 releases whose source is in https://github.com/dotnet/dotnet.

Personally, I think the name of the field shouldn't be repo-specific. I think something like this would be more appropriate:

```json
"RepoTag": "https://github.com/dotnet/dotnet/releases/tag/v8.0.100-preview.3.23178.7"
```

This same field name could then be used for 6.0/7.0 servicing builds with a field value that references `installer` instead.

### LINKED ISSUE #3382: Release pipeline: Build number should be excluded from announcement title for preview releases
The title of the release announcement that was generated by the release pipeline for the .NET 8 Preview 3 release was `.NET 8.0 Preview 3 April 2023 Update - .NET 8.0.0-preview.3.23174.8 and SDK 8.0.100-preview.3.23178.7`. It seems that we should be excluding the build number from this title. The expected title would be `.NET 8.0 Preview 3 April 2023 Update - .NET 8.0.0-preview.3 and SDK 8.0.100-preview.3`

### LINKED ISSUE #3384: Release pipeline: Allow package mirroring to be skipped
The release pipeline should allow [package mirroring](https://github.com/dotnet/source-build/blob/2c32505fda133f677d07ddc66e751e4010c11a10/eng/templates/stages/mirror.yml#L140-L165) to be skipped via a pipeline parameter. This would allow more flexibility to use the existing pipeline without modification in scenarios where there is a blocking issue related to package mirroring (see https://github.com/dotnet/source-build/issues/3381).

### PR DESCRIPTION
**Pull request by premun**
- Fixes #3384
- Fixes #3379
- Fixes #3375
- Fixes #3377
- Fixes #3376
- Fixes #3382
- Fixes #3380

> Note: Not all reported issues are issues in main, just the previous release went from a dev branch that was not sitting on top of main.

### CODE REVIEW

File: eng/templates/stages/release.yml
Comment (mthalman) [inline eng/templates/stages/release.yml:1]: It's not clear how https://github.com/dotnet/source-build/issues/3376 is fixed by these changes. Can you explain?

File: eng/source-build-release-announcement.md [hunk 1]
```diff
@@ -14,6 +14,6 @@ Please use the [$TAG tag]($TAG_URL) to source-build .NET version $RUNTIME_VERSIO
   "Channel": "$RELEASE_CHANNEL",
   "Runtime": "$RUNTIME_VERSION",
   "Sdk": "$SDK_VERSION",
-  "InstallerTag": "$TAG"
+  "RepoTag": "$TAG"
```
Comment (mthalman) [inline eng/source-build-release-announcement.md:17]: This doesn't provide context of what repo the tag is for. I'm assuming `$TAG` is still just the tag name. But it needs to include the repo with it for it to be relevant.

**Review by mthalman — COMMENTED** [review]
As a reviewer, it's difficult to review changes like this that fix a lot of posted issues. It's not easy to determine which changes are fixing which issue and how a given issue is actually fixed because it's intermingled with so many other changes.

It'd also be useful to have https://github.com/dotnet/source-build/issues/3387 implemented to verify that these changes work as intended.

File: eng/source-build-release-announcement.md (hunk 1 above)
Comment (premun) [reply eng/source-build-release-announcement.md:17]: Would we add a new property `TagUrl`?

File: eng/create-announcement-draft.sh [hunk 2]
```diff
@@ -65,23 +77,33 @@ done
...
+export RELEASE_VERSION="$release_version"
 export RUNTIME_VERSION="$runtime_version"
 export RELEASE_CHANNEL="$channel"
 export SDK_VERSION="$sdk_version"
 export RELEASE_NOTES_URL="https://github.com/dotnet/core/blob/main/release-notes/$RELEASE_CHANNEL/$RUNTIME_VERSION/$SDK_VERSION.md"
 export RELEASE_DATE=$(date +"%B %Y") # e.g. "March 2022"
 
+if [[ "$prerelease" == true ]]; then
+  # E.g. .NET 8.0 Preview 3 April 2023 Update - .NET 8.0.0-preview.3 and SDK 8.0.100-preview.3
+  export TITLE="$RELEASE_NAME $RELEASE_DATE Update - .NET $RELEASE_VERSION and SDK $RELEASE_VERSION"
```
[the remark this answers is not in the document]
Comment (premun) [reply eng/create-announcement-draft.sh:97]: Then these are new strings and we'd have to add new input parameters, probably. We don't have any idea of what part of the URL should be shown here.

Maybe we just use the gists for previews if we want something different? Or is there a heuristic?

File: eng/create-announcement-draft.sh [hunk 3]
```diff
@@ -65,23 +77,33 @@ done
...
 : "${tag:?Missing --tag}"
 
 # Set environment variables that go in the announcement template
 export TAG="$tag"
 export RELEASE_NAME="$release_name"
+export RELEASE_VERSION="$release_version"
 export RUNTIME_VERSION="$runtime_version"
 export RELEASE_CHANNEL="$channel"
 export SDK_VERSION="$sdk_version"
 export RELEASE_NOTES_URL="https://github.com/dotnet/core/blob/main/release-notes/$RELEASE_CHANNEL/$RUNTIME_VERSION/$SDK_VERSION.md"
```
[the remark this answers is not in the document]
Comment (premun) [reply eng/create-announcement-draft.sh:92]: Before, we had this:

```
./create-announcement-draft.sh" --channel "RELEASE_CHANNEL" --template...
```

Now we have this 
```
--channel "$(releaseChannel)"
```

**Comment by premun** [comment]
> As a reviewer, it's difficult to review changes like this that fix a lot of posted issues. It's not easy to determine which changes are fixing which issue and how a given issue is actually fixed because it's intermingled with so many other changes.

I agree, I originally wanted to do that but I have picked a few cosmetic ones that seemed simple enough to put together. Maybe I overdid it. I can split this up still.


> It'd also be useful to have https://github.com/dotnet/source-build/issues/3387 implemented to verify that these changes work as intended.

I imagine this one will take quite some time to implement so I wanted to get few blockers out of the way in time for the next release. I did re-run one of each releases - a servicing and a prerelease - to verify this. But to get more confidence, I believe we will have to come up with a more solid plan on how to tackle the whole pipeline. Right now it's more of a "4 pipelines quickly glued together" kind of thing.

**Comment by premun** [comment]
Closing in favour of smaller PRs

**Pull request closed without merging**